common/zeroconf/zeroconf.cpp

```diff
@@ -0,0 +1,130 @@
+                                                     0, // default all interfaces
+                                                     name.c_str(),
+                                                     "_aseba._tcp",
+                                                     NULL, // use default domain
+                                                     NULL, // use this host name
+                                                     htons(port),
+                                                     len, // TXT length
+                                                     record, // TXT record
+                                                     cb_Register,
+                                                     &z);
+        if (err != kDNSServiceErr_NoError)
+            throw runtime_error(FormatableString("Zeroconf DNSServiceRegister: error %0").arg(err));
+        else
+        {
+            DNSServiceErrorType err = DNSServiceProcessResult(z.serviceref); // block until daemon replies
+            if (err != kDNSServiceErr_NoError)
+                throw runtime_error(FormatableString("Zeroconf DNSServiceProcessResult: error %0").arg(err));
+            else
+            {
+                services[z.name] = z; // copy
+                z.serviceref = 0; // because now owned by services
+            }
+        }
+	}
+
+    //! Register and announce a target described by an existing Dashel stream
+    void Zeroconf::registerStream(const std::string name, const Dashel::Stream* stream, const string txt)
+    {
+        string service_name = name;
+        unsigned port = atoi(stream->getTargetParameter("port").c_str());
+        registerPort(name,port,txt);
+    }
+
+    //! Asynchronously update the set of known targets
+    void Zeroconf::browseForTargets()
+    {
+    }
+
+    //! callback to update ZeroconfService record with results of registration
+    void DNSSD_API Zeroconf::cb_Register(DNSServiceRef sdRef,
+                               DNSServiceFlags flags,
+                               DNSServiceErrorType errorCode,
+                               const char *name,
+                               const char *regtype,
+                               const char *domain,
+                               void *context)
+    {
+        if (errorCode != kDNSServiceErr_NoError)
+            throw runtime_error(FormatableString("Zeroconf DNSServiceRegisterReply: error %0").arg(errorCode));
+        else
+        {
+            ZeroconfService *zref = (ZeroconfService *)context;
+            zref->name = string(name);
+            zref->regtype = string(regtype);
+            zref->domain = string(domain);
+        }
+    }
+
+    //! Format target info for DNS TXT record
+    string Zeroconf::txtRecord(int protovers, std::string type, std::vector<int> ids, std::vector<int> pids)
+    {
+        std::ostringstream txt;
+        if (type.size() > 20)
+            type = type.substr(0,20);
+
+        txt.put(9);
```

Comment:
It would be safer to first put the text to be sent in a string, and then get its length, for instance:
```
string value;
...
value = "txtvers=1";
txt.put(value.length());
txt << value;